### ISSUE TITLE
Incorrect build error with linking [Blocker] 

### ISSUE THREAD

**Issue by XANOZOID** [blitz-research/monkey2#349]
# File doesn't exist in build - _but it does_

## Error Message 
```terminal
....
error: error opening './obj/local/armeabi-v7a/objs-debug/mx2_main/__/__/__/__/__/__/__/Monkey2-vGIT/monkey2/modules/stb-vorbis/stb-vorbis.buildv1.1.10/android_debug/include/_r.o.d': No such file or directory

1 error generated.

make: *** [obj/local/armeabi-v7a/objs-debug/mx2_main/__/__/__/__/__/__/__/Monkey2-vGIT/monkey2/modules/stb-vorbis/stb-vorbis.buildv1.1.10/android_debug/include/_r.o] Error 1

....
```

## Why it is happening:
1. The module which is _reportedly_ missing is not _actually_ missing
2. The code _can compile_ in one folder, but not the other
3. Shorter paths did not generate this error, _longer paths_ did

**Problem:** Long filenames are the problem. When the build process attempts to copy module files with long names, the OS fails to copy and thus this error persists. 

Semi-related to #341

**Comment by blitz-research** [comment]
Please describe how I can reproduce this, eg: where your monkey2 install is
located, what I should do to cause the error to happen etc.

An error message is nice, but without some way to reliably reproduce that
error message on my machine there's very little I can do.

**Comment by XANOZOID** [comment]
@blitz-research Sorry, no problem!

# Details
- _Monkey2 Location_: `C:\ProgFiles(x86)\Monkey2-vDEV\monkey2`
- _Project File Location_:`C:\ProgFiles(x86)\Monkey2-v1.1.08\_Projects\test_with_a_longer_name`
- _Project causing error_: Ted2Go > File > File Templates > Simple Mojo App


# Steps to reproduce:
1. You can see that the template file compiles in a directory with a short name
**example**: `C:\ProgFiles(x86)\Monkey2-v1.1.08\_Projects\test`
2. To see that somehow it breaks, just keep on making the name of the folder longer (emulating a deeper folder hierarchy) 
**example**: `C:\ProgFiles(x86)\Monkey2-v1.1.08\_Projects\test_with_a_longer_name`

3. If you've made your folder name/file hierarchy deep enough, it should generate this (semi-incorrect) error.

**Comment by blitz-research** [comment]
> To see that somehow it breaks, just keep on making the name of the
folder longer (emulating a deeper folder hierarchy)

Could you pleae post an example of such a path? Perhaps copy the URL into
the address bar of a window?

But also note that windows DOES have a finite limit on file path length of
260 chars. The NT filesystem can handle 'very long' names, but not
necessarily the OS itself so I'll never be able to make infinitely long
path names possible!  In addition 3rd party tools like emscripten, mingw,
clang, even msvc almost certainly wont be able to handle paths >260 chars
and that is totally out my control.

So this problem will always be lurking there and there's no way I can ever
completely 'solve' it. All I can really is shorten temp path names
generated by monkey2 where possible/practical, and advise people to avoid
deeeply nested folders.

Sucks, but such is life on windows. Life is even worse if you're a node.js
user: ttps://github.com/nodejs/node-v0.x-archive/issues/6960

**Comment by XANOZOID** [comment]
Sorry about the late response! This was indeed the file path which caused the linking (copying) error:

` C:\ProgFiles(x86)\Monkey2-v1.1.08\_Projects\test_with_a_longer_name\test_display.monkey2` (where test_display.monkey2 is the simple mojo app)

This was a test I made, of course. The _real_ structure was not much different, at all, actually, but I've since done some path refactoring to shorten everything as much as I can.

So that's all it really took to make the error occur, unfortunately. And, yes, I do recall having some file path length errors with Node.js fairly frequently. 

Just out of speculation, I find that it is the files being copied over to the project that is absurdly long . . . Maybe a solution would be to just make module names very short and serialized at the expense of human readability for the file names?

**Comment by blitz-research** [comment]
>  This was indeed the file path which caused the linking (copying) error:

Yes, I noticed that after my last post - will have a hack at this soon.

> I find that it is the files being copied over to the project that is
absurdly long

Which files being copied over to the project?

**Comment by XANOZOID** [comment]
Actually, it isn't _absurd_ like I wrongfully said, it's just very detailed, the file has a lot of information to it - to the point where this kind of bug happens with even relatively small file paths. 

1. The files being imported from the module have very long descriptive names
2. Debug mode adds even more context to the filename
3. The build directories add length on top of that! 

Example:
![image](https://user-images.githubusercontent.com/29788070/37139372-e14c7040-226a-11e8-9773-1d27289b9e39.png)

**Comment by blitz-research** [comment]
Latest version in develop should improve this considerably. Note you'll
need to rebuildall2go.

**Comment by XANOZOID** [comment]
Thanks, Mark! I was waiting to see what you had to say if you had finalized your solution and I have been using it since you've posted the update. The shortening code seems to be doing a great job! 

I must note that I tested to see what it would do with a project I had cloned off GitHub to my GitHub folders location and it gave some strange errors. 

This is the file path: `C:\Users\skeyt\OneDrive\Documents\GitHub\Monkey2examples`

## Error
I'll post the error below:
```console
Build error: System command failed:

cl -c -EHs -W0 -MT -utf-8 -O1 -I"C:/ProgFiles(x86)/Monkey2-vDEV/monkey2/modules/" -I"C:/ProgFiles(x86)/Monkey2-vDEV/monkey2/modules/monkey/native" -I"C:/Users/OneDrive/Documents/GitHub/Monkey2examples/" -I"C:/ProgFiles(x86)/Monkey2-vDEV/monkey2/modules/freetype/freetype-2.6.3/include/" -I"C:/ProgFiles(x86)/Monkey2-vDEV/monkey2/modules/openal/openal-soft/include/" -I"C:/ProgFiles(x86)/Monkey2-vDEV/monkey2/modules/sdl2/SDL/include/" -DBB_NEWREFLECTION -I"C:/Users/OneDrive/Documents/GitHub/Monkey2examples/Example - Bresenham Line algorithm - code example.buildv1.1.11/windows_debug_msvc_x64/build/" -showIncludes -Fo"C:/Users/OneDrive/Documents/GitHub/Monkey2examples/Example - Bresenham Line algorithm - code example.buildv1.1.11/windows_debug_msvc_x64/build/r5b58089e.cpp_r.obj" "C:/Users/OneDrive/Documents/GitHub/Monkey2examples/Example - Bresenham Line algorithm - code example.buildv1.1.11/windows_debug_msvc_x64/include/_r.cpp" >tmp/stdout1.txt

_r.cpp
C:/Users/OneDrive/Documents/GitHub/Monkey2examples/Example - Bresenham Line algorithm - code example.buildv1.1.11/windows_debug_msvc_x64/include/_r.cpp(10): fatal error C1083: Cannot open include file: 'Example - Bresenham Line algorithm - code example.buildv1.1.11/windows_debug_msvc_x64/include/Example_4_5_4Bresenham_4Line_4algorithm_4_5_4code_4example_Example_4_5_4Bresenham_4Line_4algorithm_4_5_4code_4example.h': No such file or directory

Microsoft (R) C/C++ Optimizin
```


Here's an image of what the file output looks like
![image](https://user-images.githubusercontent.com/29788070/37248014-830f041e-2483-11e8-837c-12a12ebc5779.png)


## Extra Info
[This repository](https://github.com/Pakz001/Monkey2examples) I got it from, which is pinned on the forums, has incredibly long names and I assume that a lot of people will come across this repository in hopes to learn more about the language. . . . This is obviously an extreme example and, generally, people making their own projects would not be naming their files this way at all (hopefully). But as you can see, the path length for these projects are doubled in length. Likely not to be a problem for most projects as the entry file is something under 10 or 12 letters.

What do you think? Experienced developers will be able to spot this right away and fix/report it, but newcomers (or users without lower level language experiences) may be (upset?) confused! 

## Possible next step?
If possible, I don't necessarily believe you should have to start serializing user filenames, but would it be possible to link using relative names instead of full paths? That might be the next best move here. . .

Thanks for the fixes so far, Mark!

**Comment by blitz-research** [comment]
Is this a new bug since latest fix or has it always happened?

If it's always happend, please create a new issue.

**Comment by blitz-research** [comment]
Can reproduce and am pretty sure it's a different issue, but I should be
able to do something here too. Still, those filenames are too long IMO, and
will likely give problems one way or another eventually.

**Comment by blitz-research** [comment]
Actually, it is the same issue reallly, don't bother creating a new one,
more fixes to come!

**Comment by XANOZOID** [comment]
@blitz-research Sounds good to me! I wasn't too sure myself and I can safely say that yes the error happened after the fix - unless you posted one mere minutes before you mentioned it!

**Comment by blitz-research** [comment]
Well, that's a little surprising as it should have nothing to do with the
fix!

I'll look into it thoutgh...

**Comment by XANOZOID** [comment]
Sorry, the fix you made is indeed unrelated to the error that was produced!

**Comment by blitz-research** [comment]
Wel, that's kind of  good news - I can go ahead with the v1.1.11 release
tomorrow! The fix for this will have to wait a bit though...one major
breaking change at a time!
